src/workerd/api/queue.h

```diff
@@ -63,9 +71,198 @@ class WorkerQueue: public jsg::Object {
   uint subrequestChannel;
 };
 
+// Event handler types
+
+// Types for other workers passing messages into and responses out of a queue handler.
+
+struct IncomingQueueMessage {
+	kj::String id;
+	kj::Date timestamp;
+	kj::Array<kj::byte> body;
+	JSG_STRUCT(id, timestamp, body);
+};
+
+struct QueueResponse {
+	uint16_t outcome;
+	bool retryAll;
+	bool ackAll;
+	kj::Array<kj::String> explicitRetries;
+	kj::Array<kj::String> explicitAcks;
+	JSG_STRUCT(outcome, retryAll, ackAll, explicitRetries, explicitAcks);
+};
+
+// Internal-only representation used to accumulate the results of a queue event.
+
+struct QueueEventResult {
+  bool retryAll = false;
+  bool ackAll = false;
+  kj::HashSet<kj::String> explicitRetries;
+  kj::HashSet<kj::String> explicitAcks;
+};
+
+class QueueMessage final: public jsg::Object {
+public:
+  QueueMessage(v8::Isolate* isolate, rpc::QueueMessage::Reader message, IoPtr<QueueEventResult> result);
+  QueueMessage(v8::Isolate* isolate, IncomingQueueMessage message, IoPtr<QueueEventResult> result);
+
+  kj::StringPtr getId() { return id; }
+  kj::Date getTimestamp() { return timestamp; }
+  jsg::Value getBody(v8::Isolate* isolate);
+
+  void retry();
+  void ack();
+
+  // TODO(soon): Add metadata support.
+
+  JSG_RESOURCE_TYPE(QueueMessage) {
+    JSG_READONLY_INSTANCE_PROPERTY(id, getId);
+    JSG_READONLY_INSTANCE_PROPERTY(timestamp, getTimestamp);
+    JSG_READONLY_INSTANCE_PROPERTY(body, getBody);
+    JSG_METHOD(retry);
+    JSG_METHOD(ack);
+  }
+
+private:
+  kj::String id;
+  kj::Date timestamp;
+  jsg::Value body;
+  IoPtr<QueueEventResult> result;
+
+  void visitForGc(jsg::GcVisitor& visitor) {
+    visitor.visit(body);
+  }
+};
+
+class QueueEvent final: public ExtendableEvent {
+public:
+  // TODO(cleanup): Should we get around the need for this alternative param type by just having the
+  // service worker caller provide us with capnp-serialized params?
+  struct Params {
+    kj::String queueName;
+    kj::Array<IncomingQueueMessage> messages;
+  };
+
+  explicit QueueEvent(v8::Isolate* isolate, rpc::EventDispatcher::QueueParams::Reader params, IoPtr<QueueEventResult> result);
+  explicit QueueEvent(v8::Isolate* isolate, Params params, IoPtr<QueueEventResult> result);
+
+  static jsg::Ref<QueueEvent> constructor(kj::String type) = delete;
+
+  kj::ArrayPtr<jsg::Ref<QueueMessage>> getMessages() { return messages; }
+  kj::StringPtr getQueueName() { return queueName; }
+
+  void retryAll();
+  void ackAll();
+
+  JSG_RESOURCE_TYPE(QueueEvent) {
+    JSG_INHERIT(ExtendableEvent);
+
+    JSG_READONLY_INSTANCE_PROPERTY(messages, getMessages);
+    JSG_READONLY_INSTANCE_PROPERTY(queue, getQueueName);
+
+    JSG_METHOD(retryAll);
+    JSG_METHOD(ackAll);
+  }
+
+private:
+  // TODO(perf): Should we store these in a v8 array directly rather than this intermediate kj
+  // array?
```

Comment:
I suppose it depends on how often you plan to access it. Alternatively, since you're using instance properties, you can use `JSG_READONLY_LAZY_INSTANCE_PROPERTY` such that when `getMessages()` is accessed, the conversion into a JS array will happen only once on the first call then will be cached by v8.

Reply:
Huh, neat. It'd still save some work if we just created a v8 array in the first place rather than building this array in the class's constructor, but I don't see a reason not to add that.

It's backwards-compatible to add on an existing property, right? And I assume the general guidance is only to use it on fields that are "expensive" to convert?

Reply:
Depends on your definition of backwards compatible. In certain edge cases the lazy property can be detected because it changes the own property definition from a getter to a value and that is detectable when using the reflection apis. But I believe using it here should be safe.